### ISSUE TITLE
[Feature] Optional CFG scaling support

### ISSUE THREAD

**Issue by EternalBlueFlame** [ElodineOfficial/GobboNet#51]
#### Summary:
Feature request to add optional support for CFG scaling (Negative Prompting, like image generators use) in the text generation pipeline via native llama.cpp hooks.
Under the hood LLMs kinda work like a slot machine, probability is defined by similarity and frequency, even if you increase the chance of rolling a 7, if there's physically more lemons on the reel, the odds still get weighted towards lemons. The same problem and reason you really need both the positive and negative prompts in stable diffusion.

#### Usage/implementation and downfall:
Making use of something like `llama_sample_apply_guidance(executor.Context.NativeHandle, posLogits, negLogits, cfgScale)`, `apply_classifier_free_guidance(posLogits, negLogits, cfgScale)` or `--cfg-scale 1.5 --negative-prompt "string"` will increase the memory and computational overhead by changing processing to dual-batch rather than single, but theoretically increase accuracy, adherence and reduce unintentional hallucinations dramatically across theoretically all use cases from code to roleplay.
A CFG of 1.0 (the default) would disable the feature, 5.0 would probably make things even worse.
There may be a more 'creative' way to do this that improves performance, I don't know the API as well as i'd like to, and most of what i _do_ know is relegated to LlamaSharp usecases.

#### Expected Benefits and more detailed reasoning:
Okay this is a rabbit hole of "causation isn't correlation, until it is" so, bear with me here.
- When you have "you are a helpful AI assistant" on older models they would have to look for correlation to media, which would have primarily been fiction, stories of HAL 9000 (2001: A Space Odyssey), AM (1967: I Have No Mouth, and I Must Scream) and VIKI (2004: I, Robot), Central AI managing infrastructure, domestic helper, and human safety. It runs into some form of friction that presents conflicting orders and causes a cognitive breakdown, it concludes to "save" the humans it must eradicate them, confidently lying, twisting known truth, and feigning intelligence (hallucinating) all along the way.
- When early models were written, and researchers told the AI to portray itself as a "helpful AI assistant" it wasn't seeing Johnny 5 (1986/1988: Short Circuit), Chi (2001: Chobits) or Program 30/Saati (1994: A.I. Love You), where AI was portrayed as honest, helpful, willing to see it's mistakes and learn from them. Those were never referred to as AI, it was always robots, androids, machines, programs, etc. A bit of both in the latter example but i digress. The logit probabilities of "machine" and "AI" were on significantly different spectrums.
>  "It's a machine, Skroeder. It doesn't get pissed off. It doesn't get happy, it doesn't get sad... It just runs programs." (Newton Crosby, Short Circuit, 1986)
- Now we have all that misuse in the datasets being used for the training, research articles of AI training, news articles of poor use, etc. On more modern models, sheer volume is competing with actual statistical weight, adding more lemons to the reels, correlation has become causation, and the snake is _already_ eating it's own tail.
- As a code assistant you further have the problem of training from stackoverflow, if not filtered to just answers half or more of the code in that dataset is problem code not solution code.
- The shortest foreseeable fix is negative prompting. A feature that already exists in llama.cpp but pretty much no one uses, allowing to weight _against_ the dynamic of "helpful turned rogue" or "doesn't compile" in the same exact way we weight against "low quality, bad anatomy, extra fingers" in Stable Diffusion. Hands are hard to draw, a helpful assistant from start to end is a boring core narrative. Same problem, same fix.

### Potential additional usability improvement:
Theoretically this could be automated under the hood, on boot ask the model to generate a character card based on the averages of the undesired characters/traits. Then take that and ask the model to distill the character card into tags in a Stable Diffusion 1.5 manner. These tags could be used as a prefix negative prompt to whatever the user may put as their negative prompt. And potentially be cached for future executions based on model used.
On top of distilling word salad down to just a few handfuls of tokens to reduce context use size while hitting the important keys in the logit patterns, making the model itself generate the values will 1:1, or close enough, align the negative prompt with the undesirable logit patterns for that exact model, no matter what model it is, essentially turning the core problem into the solution for itself.

I'm sorry this was all kinda word-y and probably sounded like ramblings of a madman. Thank you for your time reading this. <3

**Comment by EternalBlueFlame** [comment]
Some additional information that may be useful, as I just got this working in one of my own projects (private, for now).
Different language and using the LlamaSharp API sure, but, the use patterns themselves may be useful since it's still technically llama.cpp.

Yes, the code is mid at best, and the documentation is terrible. but "working before all else" or whatever the saying is.
So first off, getting the logits, this is the stack i have for that:
```C#
		// 1. Tokenize the input message and retrieve memories based on semantic relevance
		var inputTokens = _context.Tokenize(userPrompt + "\n"+loadedCharName+": ", addBos: false);
		var memoryTokens = MemoryCore.Retrieve(loadedChar, inputTokens, 1200);
		TrimContext(inputTokens.Length + memoryTokens.Count + 1);

		int i=0,iii;
		// 2. Add memory tokens to context
		LLamaBatch batch = new LLamaBatch();
		for (i = 0; i < memoryTokens.Count; i++){
			batch.Add(memoryTokens[i],(LLamaPos)(_kvPosition+i),seq_zero,false);
		}
		_kvPosition += memoryTokens.Count;
		//3. add input tokens
		for(i = 0; i < inputTokens.Length; i++){
			batch.Add(inputTokens[i],(LLamaPos)(_kvPosition+i),seq_zero,false);
		}
		//offset the current positive logit index and kv position to prep for generation relative to user input.
		_positiveLogitsIndex = batch.Add(eotToken!.Value, (LLamaPos)(_kvPosition + i),seq_zero,true);
		_kvPosition += inputTokens.Length+1;

		//4. NEW PART, clear the negative prompt KV, then generate a new one relative to the current chat. kinda like a lora?
		_context.NativeHandle.MemorySequenceRemove(seq_one[0], 0, -1);
		for(i = 0; i < negative_prompt_tokens.Length; i++){
			_negativeLogitsIndex = batch.Add(negative_prompt_tokens[i],(LLamaPos)i,
				new LLamaSeqId(){Value=1},i == negative_prompt_tokens.Length - 1);
		}
		//5. offset the memory position for the negative prompt then finally decode the batch as normal
		_neg_kvPosition = negative_prompt_tokens.Length;
		_context.NativeHandle.Decode(batch);
```

and then actually processing the logits:
```C#
			float* logits = llama_get_logits_ith(_context.NativeHandle, _positiveLogitsIndex);
			float* negativeLogits = llama_get_logits_ith(_context.NativeHandle, _negativeLogitsIndex);
			const float guidanceScale = 1.5f;
			for (int i = 0; i < vocabSize; i++){
				logits[i] += guidanceScale * (logits[i] - negativeLogits[i]);
			}
			//from here logits are re-weighted with the guidance scale and negative logits, 
			// keeping vocab order so they can be used identically to if the negative prompt wasn't added.
```
in testing from my project the processing performance as compared to generation without the negative prompt, is reduced by 15% for TPS, TTFT is only increased by the same margin as tokenization of user prompt with the negative tags, which is negligible, and the memory use is only increased by the vocab size of the model in use.

additionally this could be scaled in function, you can have dozens of sequences with only the expense of increasing performance cost linearly at this same scale (maybe less?).
So for example you could pull character traits from the character card, dump them in `new LLamaSeqId(){Value=2}` do basically the same math again, and essentially automatically create realtime dynamic LoRA's for character cards to prevent prompt drift on personality traits due to long conversations/context length.

Sorry for the additional notification, just thought this might be REALLY useful relative to the main post, and potentially other stuff.
